cpp/src_prims/metrics/trustworthiness_score.cuh

```diff
@@ -27,41 +26,38 @@ namespace Score {
 
 /**
      * @brief Compute a the rank of trustworthiness score
-     * @param[in] X_ind: Indexes of pairwise distance calculations of X and sorting
-     * @param[in] X_dist: Distances of pairwise distance calculations of X and sorting
+     * @param[in] X_ind: Sorted indexes of pairwise distance calculations of X
+     * @param[in] X_nn_order: Nearest neighbors order
+  *                            of pairwise distance calculations of X
      * @param[in] emb_ind: Indexes of KNN on embeddings
      * @param[in] emb_dist: Distances of KNN on embeddings
      * @param n: Number of samples
      * @param n_neighbors: Number of neighbors considered by trustworthiness score
      * @param work: Batch to consider (to do it at once use n * n_neighbors)
      * @param[out] rank: Resulting rank
      */
-template <typename math_t, typename knn_index_t>
-__global__ void compute_rank(int *X_ind, math_t *X_dist, knn_index_t *emb_ind,
-                             math_t *emb_dist, int n, int n_neighbors, int work,
-                             int offset, double *rank) {
+template <typename knn_index_t>
+__global__ void compute_rank(int *X_ind, int *X_nn_order, knn_index_t *emb_ind,
+                             int n, int n_neighbors, int work, int offset,
+                             double *rank) {
   int i = blockIdx.x * blockDim.x + threadIdx.x;
   if (i >= work) return;
 
   int sample_idx = i / n_neighbors;
   int nn_idx = i % n_neighbors;
 
-  int loc = sample_idx * n_neighbors + nn_idx;
-  knn_index_t emb_nn_ind = emb_ind[loc];
+  knn_index_t emb_nn_ind = emb_ind[sample_idx * n_neighbors + nn_idx];
 
   if ((offset + nn_idx) == emb_nn_ind) return;
 
-  math_t emb_nn_dist = emb_dist[loc];
-
-  loc = sample_idx * n;
+  int loc = sample_idx * n;
   int *X_pw_ind = &X_ind[loc];
-  math_t *X_pw_dist = &X_dist[loc];
 
   knn_index_t left = 0;
   knn_index_t right = n - 1;
   while (left < right) {
     knn_index_t mid = std::floor((left + right) / 2);
-    if (X_pw_dist[mid] < emb_nn_dist) {
+    if (X_pw_ind[mid] < emb_nn_ind) {
```

Comment:
Does this assume that the pairwise distances are sorted by index?

Reply:
Ah, I think I see- you have created an additional lookup table to get the distance based on the sort position. If the binary search is showing worse performance even on worse datasets, just thinking briefly about your idea here, I'm wondering if it would just be worth taking this idea a step further and pre-building a lookup table that directly maps (knn row, k) -> position in resulting pw matrix. That way you wouldn't even need to compute / keep the distances at all and the rank computation just becomes a coalesced element-wise op.